---
title: "MSA - Center Star Algorithm"
author: "Cy Chittenden"
date: "2/20/2022"
output: html_document
---

```{r}
suppressMessages(library(Biostrings)) #silently loads Biostrings functions
source("pairAlign.R")

```

#buildMSA function provided
```{r}

buildMSA <- function(patterns, subjects, center) {
    MSA <- rbind(patterns[[1]], subjects[[1]])
    for(i in 2:length(patterns)) {
        j = 1 #index in new row
        k = 1 #index in alignment of center sequence to sequence i
        m = 1 #column index of MSA
        maxLength = ncol(MSA) + length(patterns[[i]])
        newRow = character(maxLength)
        while(k <= length(patterns[[i]]) && m <= ncol(MSA)) {
            if(patterns[[i]][k] == MSA[1, m]) {
                newRow[j] <- subjects[[i]][k]
                j <- j + 1
                k <- k + 1
                m <- m + 1
            } else if(MSA[1, m] == "-") {
                newRow[j] <- "-"
                j <- j + 1
                m <- m + 1
            } else if(patterns[[i]][k] == "-") {
                if(m == 1) {
                    MSA <- cbind(rep("-", nrow(MSA)), MSA)
                } else {
                    MSA <- cbind(MSA[, 1:(m-1)], rep("-", nrow(MSA)), MSA[, m:ncol(MSA)])
                }
                newRow[j] <- subjects[[i]][k]
                m <- m + 1
                j <- j + 1
                k <- k + 1
            }
        }
        while(k <= length(patterns[[i]])) {
            MSA <- cbind(MSA, rep("-", nrow(MSA)))
            newRow[j] <- subjects[[i]][k]
            k <- k + 1
            j <- j + 1
        }
        while(m <= ncol(MSA)) {
            newRow[j] <- "-"
            m <- m + 1
            j <- j + 1
        }
        newRow <- newRow[1:(m - 1)]
        MSA <- rbind(MSA, newRow)
    }
    rownames(MSA) <- c("    Center:", paste0("Sequence ", 1:(nrow(MSA)), ":")[-center])
    colnames(MSA) <- 1:ncol(MSA)
    return(MSA)
}
```



```{r}
centerStar <- function(sequences, substitutionMatrix, gapOpening, gapExtension) {
    
    n <- length(sequences)
    alignmentPatterns <- list()
    alignmentSubjects <- list()
    k <- 1

    for(i in 1:(length(sequences) -1)) { #for each sequence except the last one
        for(j in (i+1):length(sequences)) { #for each j larger than i in sequences
            aligned <- pairAlign(sequences[i], sequences[j], substitutionMatrix = substitutionMatrix, type = "global", gapOpening = gapOpening, gapExtension = gapExtension)#use pairAlign() here to find the pattern and subject
            alignmentPatterns <-  cbind(alignmentPatterns,aligned[1]) #store the alignment sequence of the first protein in alignmentPatterns
            alignmentSubjects <- cbind(alignmentSubjects,aligned[2]) #store the alignment sequence of the second protein in alignmentSubjects
       
        }
    }
    
    
    
    distVector <- numeric(length(alignmentPatterns)) #create distVector, a vector that's the length of alignmentPatterns
    for (i in 1:length(alignmentPatterns)) {
    pattern <-strsplit(alignmentPatterns[[i]], split = "")[[1]]
    subject <- strsplit(alignmentSubjects[[i]], split = "")[[1]]
    distance <- pattern != subject
    distVector[i] <-  sum(distance) #split the two sequences to be aligned into multiple strings of amino acid characters, and count the number of mismatches between them
    }
    distVector #great, now you have your vector with the distances inside it

#use code from above to make a matrix that looks like exMatrix, but with your distVector instead

    distMatrix = matrix(0, nrow = n, ncol = n)
    distMatrix[lower.tri(distMatrix)] <- distVector
    distMatrix <- distMatrix + t(distMatrix)
    
    
    
    
    exSums <- colSums(distMatrix) #sum columns in distMatrix
    exCandidates <- which(exSums == min(exSums)) #returns which column has the lowest colSums value
    center <- exCandidates[nnet::which.is.max(nchar(sequences[exCandidates]))]
    
    
    
    
    n <- length(sequences)
    a <- length(alignmentPatterns)

#create a list to store our patterns and our subjects
    PATTERNS <- list()
    SUBJECTS <- list()
#create our 2 matrices and fill them with "-"s
    subpat = matrix("-", nrow = n, ncol = n)
    patsub = matrix("-", nrow = n, ncol = n)

    alignmentPatternsV <- c()
    for (i in 1:a){
    alignmentPatternsV[i] <- alignmentPatterns[[i]]
    } #store elements in alignmentPatterns, a list, in alignPatternsV, a vector
    alignmentSubjectsV <- c()
    for (i in 1:a){
    alignmentSubjectsV[i] <- alignmentSubjects[[i]]
    } #store elements in alignmentSubjects, a list, in alignSubjectsV, a vector


 #fill lower triangle of subpat with alignmentPatternsV
    subpat[lower.tri(subpat)] <- alignmentPatternsV

 #fill lower triangle of patsub with alignmentSubjectsV
    patsub[lower.tri(patsub)] <- alignmentSubjectsV

 #fill upper triangle of subpat with transposed lower triangle of patsub
    subpat[upper.tri(subpat)] <- t(patsub)[upper.tri(patsub)]


 #fill upper triangle of patsub with transposed lower triangle of subpat
    patsub[upper.tri(patsub)] <- t(subpat)[upper.tri(subpat)]

#subpat now contains subjects as rows and patterns as columns
#patsub now contains patterns as rows and subjects as columns

    PATTERNS <-  strsplit(subpat[,center][subpat[,center] != "-"], split = "")#store patterns, split into single character strings, in PATTERNS
    SUBJECTS <-  strsplit(patsub[,center][patsub[,center] != "-"], split = "")#store subjects, split into single character strings, in SUBJECTS


    
    
    return(buildMSA(PATTERNS, SUBJECTS, center))
}

globins <- unlist(seqinr::read.fasta("globins.fasta", seqtype = "AA", as.string = T))

centerStar(sequences = globins, substitutionMatrix = "BLOSUM62", gapOpening = -10, gapExtension = -2)
```